conmon-rs/server/src/oom_watcher.rs

```diff
@@ -0,0 +1,270 @@
+        let mut last_counter: u64 = 0;
+
+        let (mut watcher, mut rx) = Self::async_watcher()?;
+        debug!(
+            "{}:watching oom path {}",
+            pid,
+            memory_events_file_path.display()
+        );
+        watcher.watch(&memory_events_file_path, RecursiveMode::Recursive)?;
+
+        loop {
+            tokio::select! {
+                _ = exit_rx.recv() => {
+                    break;
+                }
+                Some(res) = rx.recv() => {
+                    match res {
+                        Ok(event) => {
+                            if event.kind.is_remove() {
+                                break
+                            }
+                            if !event.kind.is_modify() {
+                                continue;
+                            }
+                             debug!("{}: found modify event", pid);
+                            let (counter, is_oom) = Self::check_for_oom(&memory_events_file_path, last_counter).await?;
+                            if is_oom {
+                                debug!("{}: found oom event with counter {}", pid, counter);
+                                if let Err(e) = Self::write_oom_files(exit_paths).await {
+                                    error!("error writing oom files: {}", e);
+                                }
+                                last_counter = counter;
+                            }
+                        },
+                        Err(e) => {
+                            debug!("watch error: {:?}", e);
+                        },
+                    };
+                }
+            }
+        }
+        watcher.unwatch(&memory_events_file_path)?;
+
+        debug!("{}: done watching for ooms", pid);
+
+        Ok(())
+    }
+
+    async fn check_for_oom(
+        memory_events_file_path: &Path,
+        last_counter: u64,
+    ) -> Result<(u64, bool)> {
+        let mut new_counter: u64 = 0;
+        let mut found_oom = false;
+        let fp = File::open(memory_events_file_path).await?;
+        let reader = BufReader::new(fp);
+        let mut lines = reader.lines();
+        while let Some(line) = lines.next_line().await? {
+            debug!("{}", line);
+            if let Some(counter) = line.strip_prefix("oom ") {
+                let counter = counter.to_string().parse::<u64>()?;
+                if counter != last_counter {
+                    new_counter = counter;
+                    found_oom = true;
+                    break;
+                }
+            }
+        }
+        Ok((new_counter, found_oom))
+    }
+
+    async fn write_oom_files(exit_paths: &[PathBuf]) -> Result<()> {
+        let paths = exit_paths.to_owned();
+        let tasks: Vec<_> = paths
+            .into_iter()
+            .map(|path| {
+                tokio::spawn(async move {
+                    let fp = File::create(path.join("oom")).await;
```

Comment:
instead of joining the path, I think we should just write directly to the path, so there isn't less secret knowlege the caller and server need to share

Reply:
fixed.